File: 0092-Reverse_Linked_List_II/Reverse_Linked_List_II.java
class Solution {
    ListNode nodeAfterReversedEnd;
    ListNode reversedStart;
    
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if(left == right) return head;  // No change is required
        
        ListNode temp = head;   // If the first node needs to be reverse, create a placeHolder previous node
        if(left == 1) {
            temp = new ListNode();
            temp.next = head;
        }
        
        ListNode newHead = temp;    
        for(int i = 1; i < left - 1; i++){
            temp = temp.next;       // Traverse to the node before the swap take place
        }
        
        ListNode reverseEnd = reverseNodes(temp.next, (right - left));  // Get the last node after reverse
        reverseEnd.next = nodeAfterReversedEnd; 
        temp.next = reversedStart; 
        
        return left == 1 ? newHead.next : newHead;
    }
    
    // method to reverse
    private ListNode reverseNodes(ListNode node, int count){
        if(count == 0){
            nodeAfterReversedEnd = node.next;
            reversedStart = node;
            return node;
        }
        count--;
        reverseNodes(node.next, count).next = node;

        return node;
    }
}

File: 0092-ReverseLinkedListII/ReverseLinkedListII.go
package reverselinkedlistii

type ListNode struct {
	Val  int
	Next *ListNode
}

var nodeAfterReversedEnd *ListNode
var reversedStart *ListNode

func reverseBetween(head *ListNode, left int, right int) *ListNode {
	if left == right {
		return head
	}

	temp := head
	if left == 1 {
		temp = &ListNode{Next: head}
	}

	newHead := temp
	for i := 1; i < left-1; i++ {
		temp = temp.Next
	}

	reverseEnd := reverseNodes(temp.Next, right-left)
	reverseEnd.Next = nodeAfterReversedEnd
	temp.Next = reversedStart

	if left == 1 {
		return newHead.Next
	}

	return newHead
}

func reverseNodes(node *ListNode, count int) *ListNode {
	if count == 0 {
		nodeAfterReversedEnd = node.Next
		reversedStart = node
		return node
	}

	count--
	reverseNodes(node.Next, count).Next = node
	return node
}
